File: solutions/aoc2021_day06.js
var fs = require('fs');

function count_fish(pi, days) {
    var initial = pi.split(',').map(Number);
    var count = {};
    for (var fish of initial) {
        if (count[fish] === undefined) count[fish] = 0;
        count[fish] += 1;
    }
    for (var day = 0; day < days; day++) {
        var new_count = {};
        for (const k in count) {
            const v = count[k];
            if (k > 0) {
                if (new_count[k-1] === undefined) new_count[k-1] = 0;
                new_count[k-1] += v;
            }
            else {
                if (new_count[6] === undefined) new_count[6] = 0;
                if (new_count[8] === undefined) new_count[8] = 0;
                new_count[6] += v;
                new_count[8] += v;
            }
        }
        count = new_count;
    }
    return Object.values(count).reduce(function(a, b) { return a + b; });
}

function part1(pi) {
    var ans = count_fish(pi, 80);
    console.log(`part1: ${ans}`);
}

function part2(pi) {
    var ans = count_fish(pi, 256);
    console.log(`part2: ${ans}`);
}

console.log("day06");
var f = fs.readFileSync('aoc2021_day06.txt', 'utf8');
var pi = f.toString();
part1(pi);
part2(pi);

File: solutions/aoc2021_day06.py
import copy

def part1(arr):
    fishes = [int(e) for e in arr[0].split(',')]
    pop = copy.deepcopy(fishes)
    for _ in range(80):
        n = len(pop)
        for i in range(n):
            if pop[i] > 0:
                pop[i] -= 1
            else:
                pop[i] = 6
                pop.append(8)
    ans = len(pop)
    print(ans)

def part2(arr):
    fishes = [int(e) for e in arr[0].split(',')]
    count = dict((fish, 0) for fish in fishes)
    for fish in fishes: count[fish] += 1
    for _ in range(256):
        new_count = {}
        for k, v in count.items():
            if k > 0:
                if k-1 not in new_count: new_count[k-1] = 0
                new_count[k-1] += v
            else:
                if 6 not in new_count: new_count[6] = 0
                if 8 not in new_count: new_count[8] = 0
                new_count[6] += v
                new_count[8] += v
        count = new_count
    ans = sum(count.values())
    print(ans)

with open("aoc2021_day06.txt") as f:
    data = f.read()
    arr = [line for line in data.splitlines()]

print("day06")
part1(arr)
part2(arr)
